stl/src/filesystem.cpp

```diff
@@ -665,6 +675,51 @@ _Success_(return == __std_win_error::_Success) __std_win_error
 
 [[nodiscard]] __std_win_error __stdcall __std_fs_rename(
     _In_z_ const wchar_t* const _Source, _In_z_ const wchar_t* const _Target) noexcept {
+    __std_win_error _Last_error;
+    constexpr auto _Flags = __std_fs_file_flags::_Backup_semantics | __std_fs_file_flags::_Open_reparse_point;
+    // `MoveFileExW` requires that the user has 'DELETE' access to the files
+    constexpr auto _Common_access_rights = __std_access_rights::_Delete | __std_access_rights::_File_read_attributes;
+
+    { // dummy scope so that the destructors of _Fs_file can be called
+        _STD _Fs_file _Target_handle(_Target, _Common_access_rights, _Flags, &_Last_error);
+
+        if (_Last_error == __std_win_error::_Success) {
+
+            _STD _Fs_file _Source_handle(_Source, _Common_access_rights, _Flags, &_Last_error);
+            if (_Last_error == __std_win_error::_Success) {
+                if (_Is_same_file(_Source_handle._Get(), _Target_handle._Get())) {
+                    // Renaming files which are hardlinks of each other is a no-op.
+                    return __std_win_error::_Success;
+                }
+            } else if (_Last_error == __std_win_error::_File_not_found) {
+                return _Last_error;
+            }
+
+            // Choosing FILE_STANDARD_INFO over FILE_BASIC_INFO to check whether the handles refer to directories.
+            // Because sizeof(FILE_STANDARD_INFO) = 24 < sizeof(FILE_BASIC_INFO) = 40.
+            // How expensive is this operation? (comment for reviewers)

```

Comment:
My guess it does not matter much. Though `FILE_STANDARD_INFO` is more demanded information while working with file, so more likely to be easier to get.